Mathlib/LinearAlgebra/Matrix/FixedDetMatrices.lean

```diff
@@ -52,4 +53,261 @@ lemma smul_coe (m : R) (g : SpecialLinearGroup n R) (A : FixedDetMatrix n R m) :
+            ⟨M.1.1 1 1, reps_entries_le_m' k hk M M.2 1 1⟩⟩)
+    intro M N h
+    ext i j
+    rw [Prod.ext_iff] at h
+    simp only [Fin.isValue, Subtype.mk.injEq, Prod.mk.injEq] at h
+    fin_cases i <;> fin_cases j --is there an easy way to golf this?
+    · exact h.1
+    · exact h.2.1
+    · exact h.2.2.1
+    · exact h.2.2.2
+  · simp only [ne_eq, Decidable.not_not] at hk
+    rw [hk, reps_zero_empty]
+    exact Set.fintypeEmpty
+
+lemma reduce_mem_reps (m : ℤ) (hm : m ≠ 0) : ∀ A : Δ m, reduce m A ∈ reps m := by
+  apply reduce_rec
+  · intro A h
+    by_cases h1 : 0 < A.1 0 0
+    · simp only [reps, Fin.isValue, reduce_eqn1 m A h h1, zpow_neg, Set.mem_setOf_eq, smul_coe,
+      coe_inv, coe_T_zpow, adjugate_fin_two_of, neg_zero, cons_mul, Nat.succ_eq_add_one,
+      Nat.reduceAdd, empty_mul, Equiv.symm_apply_apply, of_apply, cons_val', vecMul,
+      cons_dotProduct, vecHead, one_mul, vecTail, Function.comp_apply, Fin.succ_zero_eq_one,
+      neg_mul, dotProduct_empty, add_zero, zero_mul, zero_add, empty_val', cons_val_fin_one,
+      cons_val_one, cons_val_zero, lt_add_neg_iff_add_lt, le_add_neg_iff_add_le]
+      refine ⟨abs_eq_zero.mp h, ?_, ?_, ?_⟩
+      · rw [abs_eq_zero.mp h]; simp only [Fin.isValue, mul_zero, h1]
+      · apply Int.ediv_mul_le; apply A_d_ne_zero m A (by simpa only [abs_eq_zero] using h) hm
+      · rw [mul_comm, ← @Int.sub_eq_add_neg, (Int.emod_def (A.1 0 1) (A.1 1 1)).symm]
+        apply le_trans _ (Int.emod_lt (A.1 0 1) (A_d_ne_zero m A (by simpa using h) hm))
+        rw [abs_eq_self.mpr (Int.emod_nonneg (A.1 0 1) (A_d_ne_zero m A (by simpa using h) hm))]
+    · simp only [reps, Fin.isValue, reduce_eqn2 m A h h1, Int.ediv_neg, neg_neg, smul_def, ←
+      mul_assoc, S_mul_S_eq, neg_mul, one_mul, coe_T_zpow, mul_neg, cons_mul, Nat.succ_eq_add_one,
+      Nat.reduceAdd, empty_mul, Equiv.symm_apply_apply, neg_of, neg_cons, neg_empty,
+      Set.mem_setOf_eq, of_apply, cons_val', Pi.neg_apply, vecMul, cons_dotProduct, vecHead,
+      vecTail, Function.comp_apply, Fin.succ_zero_eq_one, dotProduct_empty, add_zero, neg_add_rev,
+      zero_mul, zero_add, empty_val', cons_val_fin_one, cons_val_one, neg_eq_zero, cons_val_zero,
+      lt_add_neg_iff_add_lt, le_add_neg_iff_add_le, abs_neg]
+      refine ⟨abs_eq_zero.mp h, ?_, ?_, ?_⟩
+      · simp only [Fin.isValue, abs_eq_zero.mp h, mul_zero, neg_zero, Int.lt_iff_le_and_ne, ne_eq]
+        refine ⟨not_lt.mp h1, A_a_ne_zero m A (by simpa using h) hm⟩
+      · rw [le_neg]
+        apply Int.ediv_mul_le; apply A_d_ne_zero m A (by simpa using h) hm
+      · rw [mul_comm, add_comm, ← @Int.sub_eq_add_neg, (Int.emod_def (-A.1 0 1) (A.1 1 1)).symm]
+        apply le_trans _ (Int.emod_lt (-A.1 0 1) (A_d_ne_zero m A (by simpa using h) hm))
+        rw [abs_eq_self.mpr (Int.emod_nonneg (-A.1 0 1) (A_d_ne_zero m A (by simpa using h) hm))]
+  · exact fun A h1 h2 ↦ Eq.mpr (id (congrArg (fun _a ↦ _a ∈ reps m) (reduce_eqn3 m A h1))) h2
```

Comment:
```suggestion
  · exact fun A h1 h2 ↦ reduce_eqn3 m A h1 ▸ h2
```
You should not use the automatically generated lemma `reduce_eqn3` by name. It is better to state the result explicitly (with a suitable name) and prove it by `rfl`, so that it can then be used with a reasonable name.